bindings/protocols/coap/index.html

```diff
@@ -1,181 +1,464 @@
-            <td>CoAP method name (Literal).</td>
-            <td>optional</td>
-            <td>
-              <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a>
-              <p>(one of <code>"GET"</code> (1), <code>"POST"</code> (2), <code>"PUT"</code> (3), <code>"DELETE"</code> (4),
-                <code>"FETCH"</code> (5), <code>"PATCH"</code> (6), <code>"iPATCH"</code> (7))</p>
-            </td>
-          </tr>
-          <tr>
-            <td><code>cov:options</code></td>
-            <td>CoAP options sent with the message, e.g., <code>[ { "cov:optionName": "Accept", "cov:optionValue": 110 }]</code> to observe.</td>
-            <td>optional</td>
-            <td>
-              array of <code>cov:MessageOption</code>
-            </td>
-          </tr>
-          <tr>
-            <td><code>cov:optionName</code></td>
-            <td>Option name (Literal), see <a
-                href="https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#option-names">CoRE
-                Parameters</a>.</td>
-            <td>mandatory within <code>cov:MessageOption</code></td>
-            <td>
-              <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a>
-            </td>
-          </tr>
-          <tr>
-            <td><code>cov:optionValue</code></td>
-            <td>Header value (Literal).</td>
-            <td>mandatory within <code>cov:MessageOption</code></td>
-            <td>
-              <a href="https://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#anyType"><code>anyType</code></a>
-            </td>
-          </tr>
-        </tbody>
-      </table>
+    <section id="overview">
+        <h2>Protocol Overview</h2>
+        <p class="ednote">TODO: Provide an overview of the protocol and the protocol binding.</p>
+        <table class="def">
+            <thead>
+                <tr>
+                    <th>Feature</th>
+                    <th>Covered by</th>
+                </tr>
+            </thead>
+            <tbody>
+                <tr>
+                    <td>Request to Server</td>
+                    <td><code>"href": ..., "coap:method": ...</code></td>
+                </tr>
+                <tr>
+                    <td>Request to Forward Proxy</td>
+                    <td><code>"href": ..., "coap:method": ..., "coap:proxy": ...</code></td>
+                </tr>
+                <tr>
+                    <td>Caching: Validation</td>
+                    <td>n/a</td>
+                </tr>
+                <tr>
+                    <td>Conditional Requests</td>
+                    <td>n/a</td>
+                </tr>
+                <tr>
+                    <td>Content</td>
+                    <td><code>"coap:format": ...</code></td>
+                </tr>
+                <tr>
+                    <td>Content Negotiation</td>
+                    <td><code>"coap:accept": ...</code></td>
```

Comment:
As discussed in #49, `coap:format` and `coap:accept` could also be implied by the `contentType` and/or the [`response`](https://w3c.github.io/wot-thing-description/#td-vocab-response--Form) fields.

Reply:
True. I think we need to discuss our assumptions regarding a client's knowledge of content types and CoAP content formats: Do we assume that a client is capable of translating a string such as "text/plain; charset=utf-8" to the ID 0 (which would pretty much require that the client has a copy of the IANA registry) or would it be easier if the ID was included in the TD right away?

Reply:
Re-reading [the section on `Form`](https://w3c.github.io/wot-thing-description/#form), it's also a bit unclear to me what `contentType` and `contentCoding` actually indicate. Do these describe the format of the payload that the client is supposed to submit (&rarr; Content-Format option in CoAP request), the format of the response payload to expect (&rarr; Content-Format option in CoAP response), or the format of the response payload that the client is supposed to request (&rarr; Accept option in CoAP request)?

Reply:
I found that a bit unclear as well. My latest understanding was that `contentType` describes the payload format both of requests and responses unless there is a response field set which only refers to the response. The spec also states:

> The optional response name-value pair can be used to provide metadata for the expected response message. [...] If no response name-value pair is provided, it MUST be assumed that the content type of the response is equal to the content type assigned to the Form instance.

So if both are set, I think you could argue that `contentType` should refer to the Content-Format option and `response.contentType` should refer to the `Accept` option. If only the Form `contentType` is given, then it would refer to both options. I am not sure if this is the correct interpretation. Maybe some clarification in the TD spec would be required. 

node-wot is currently using the contentType only for the acccept option by the way: https://github.com/eclipse/thingweb.node-wot/blob/master/packages/binding-coap/src/coap-client.ts#L256 The Content-Format needs to be explicitly set as an option in the Form.

> I think we need to discuss our assumptions regarding a client's knowledge of content types and CoAP content formats: Do we assume that a client is capable of translating a string such as "text/plain; charset=utf-8" to the numeric ID 0 (which would pretty much require that the client has a copy of the IANA registry) or would it be easier if the numeric ID was included in the TD right away?

That is a very good question! I think the mapping between string-based contentTypes and CoAP's Content-Format is also a bit difficult in general if you have (additional) parameters. Being able to use the value directly is certainly better, especially for consumers that are more constrained. Maybe for TD 2.0 it could also be allowed to use an integer for `contentType`? In this case the protocol specific meaning could be inferred from the `href`'s URI scheme. Or would it be better to keep it only for human-readable information?

In general, I think there should probably be some kind of recommendation/(pseudo-)assertion that the values have to match as given by the IANA registry if both `coap:format` and `contentType` are being used.

Reply:
> I think we need to discuss our assumptions regarding a client's knowledge of content types and CoAP content formats: Do we assume that a client is capable of translating a string such as "text/plain; charset=utf-8" to the numeric ID 0 (which would pretty much require that the client has a copy of the IANA registry) or would it be easier if the numeric ID was included in the TD right away?

This was also a discussion for MQTT where we could have used the integer values that are actually put into the header instead of `"SUBSCRIBE"` for example. I have then looked at some libraries and none of them actually expose these integer values to the programming API. Thus, we have chosen the words instead of the numbers. Here I would recommend the same thing: Go the way most libraries' APIs have gone